### PR TITLE
Initial iteration of gradient boosting

### PR DESCRIPTION
**Pull request by flux-wrk**
Code needs some refactoring. as well as readme with usage instruction, but it will do as first iteration.

### CODE REVIEW

File: gradientboost/libs/desicion_tree/BoostedClassifier.cpp [hunk 1]
```diff
@@ -0,0 +1,107 @@
+#include "BoostedClassifier.h"
+
+NGradientBoost::BoostedClassifier::DecisionTree::DecisionTree(size_t depth) : depth_(depth) {
+    splitting_features_ = std::vector<size_t>();
```
Comment (messteno) [inline gradientboost/libs/desicion_tree/BoostedClassifier.cpp:4]: Вектор не надо инициализировать.

File: gradientboost/libs/desicion_tree/BoostedClassifier.h [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+#include <string>
+#include <set>
+#include <limits>
+#include "DataFrame.h"
+
+namespace NGradientBoost {
+
+    class BoostedClassifier {
+    private:
+        class DecisionTree {
```
Comment (messteno) [inline gradientboost/libs/desicion_tree/BoostedClassifier.h:11]: Я бы сделал всё же классом снаружи, и при необходимости подружил их, но это место я и так переделаю

File: gradientboost/libs/desicion_tree/BoostedClassifier.cpp [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    leaf_answers_ = std::vector<float_t>(static_cast<unsigned long>(1 << (depth_ + 1)), 0.0);
```
Comment (messteno) [inline gradientboost/libs/desicion_tree/BoostedClassifier.cpp:5]: Лучше по возможности все переменные инициализировать вне тела конструктора

File: gradientboost/libs/desicion_tree/BoostedClassifier.cpp [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        for (int j = 0; j < depth_; ++j) {
+            mask += (dataframe[i][splitting_features_[j]] << (depth_ - j));
+        }
+        res[i] = leaf_answers_[mask];
+    }
+    return res;
+}
+
+NGradientBoost::BoostedClassifier &
+NGradientBoost::BoostedClassifier::Fit(const std::vector<std::vector<float_t>> &data, const std::vector<float_t> target) {
```
Comment (messteno) [inline gradientboost/libs/desicion_tree/BoostedClassifier.cpp:23]: std::vector<std::vector<float_t>> очень длинные конструкции, лучше заменить на какой-нибудь
`
using FeaturesRow = std::vector<float_t>;
using DataSet = std::vector<FeaturesRow>;
`

File: gradientboost/libs/desicion_tree/BoostedClassifier.cpp [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    auto dataframe = DataFrame(data);
+    trees_.clear();
+    std::vector<float_t> current_predictions(dataframe.size(), 0), temp_pred(dataframe.size(), 0);
+
+    for (size_t iteration = 0; iteration < tree_count_; ++iteration) {
+        DecisionTree weak_classifier(tree_depth_);
+        std::vector<int> leaf_indices(dataframe.size(), 0);
+
+        for (size_t depth = 0; depth < tree_depth_; ++depth) {
```
Comment (messteno) [inline gradientboost/libs/desicion_tree/BoostedClassifier.cpp:32]: Этот кусок точно надо будет разбить на несколько функций

File: gradientboost/libs/desicion_tree/BoostedClassifier.cpp [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+            used_features.insert(best_feature / dataframe.get_bin_count());
+            leaf_indices = best_leaf_ind;
+
+            for (size_t i = 0; i < dataframe.size(); ++i) {
+                temp_pred[i] = current_predictions[i] + best_leaf_ans[best_leaf_ind[i]];
+            }
+        }
+        current_predictions = temp_pred;
+
+        float_t MSE = 0.0;
```
Comment (messteno) [inline gradientboost/libs/desicion_tree/BoostedClassifier.cpp:85]: Тоже в отдельную функцию бы вынести

File: gradientboost/libs/desicion_tree/BoostedClassifier.h [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+            std::vector<float_t> Predict(const std::vector<std::vector<float_t>>& data) const;
+        };
+
+        std::vector<DecisionTree> trees_;
+        float_t learning_rate_{};
+        size_t tree_count_;
+        size_t tree_depth_;
+
+    public:
+        BoostedClassifier(float_t learning_rate, size_t depth, size_t tree_count) {
```
Comment (messteno) [inline gradientboost/libs/desicion_tree/BoostedClassifier.h:28]: Также лучше инициализировать до тела конструктора

File: gradientboost/libs/desicion_tree/BoostedClassifier.h [hunk 8]
```diff
@@ -0,0 +1,39 @@
+#include <vector>
+#include <string>
+#include <set>
+#include <limits>
+#include "DataFrame.h"
+
+namespace NGradientBoost {
+
+    class BoostedClassifier {
```
Comment (messteno) [inline gradientboost/libs/desicion_tree/BoostedClassifier.h:9]: Для бустинга вообще бы лучше отдельную библиотеку сделать

File: gradientboost/libs/preprocessing/Preprocessor.h [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+#include <cfloat>
+#include <cmath>
+#include <algorithm>
+#include <memory>
+
+#include <set>
+#include <map>
+
+
+namespace NGradientBoost {
```
Comment (messteno) [inline gradientboost/libs/preprocessing/Preprocessor.h:15]: Предлагаю не делать отступ внутри namespace, а то код будет сильно разъезжаться

File: gradientboost/libs/preprocessing/Preprocessor.h [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    template <typename TInput, typename TOutput>
+    class Transformer {
+     public:
+        virtual Transformer<TInput, TOutput>& fit(const std::vector<std::vector<TInput>>& dataframe, size_t column) = 0;
+        virtual void transform(const TInput& data, std::vector<TOutput>& output) = 0;
```
Comment (messteno) [inline gradientboost/libs/preprocessing/Preprocessor.h:20]: Методы классов в разных местах то с большой буквы, то с маленькой, давай везде с большой.
